compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs

```diff
@@ -119,24 +120,50 @@ generateSrcPkgFromLf thisPkgId pkgMap pkg = do
             toNormalizedFilePath $
             (joinPath $ map T.unpack $ LF.unModuleName $ LF.moduleName mod) <.>
             ".daml"
-    pure ( fp
-         , unlines header ++
-           (showSDocForUser fakeDynFlags alwaysQualify $
-            ppr $ generateSrcFromLf thisPkgId pkgMap mod))
+    pure
+        ( fp
+        , unlines header ++
+          (showSDocForUser fakeDynFlags alwaysQualify $
+           ppr $ generateSrcFromLf (DontQualify False) thisPkgId pkgMap mod) ++
+          unlines (builtins mod))
   where
-    header = ["{-# LANGUAGE NoDamlSyntax #-}", "{-# LANGUAGE NoImplicitPrelude #-}"]
+    header =
+        [ "{-# LANGUAGE NoDamlSyntax #-}"
+        , "{-# LANGUAGE NoImplicitPrelude #-}"
+        , "{-# LANGUAGE TypeOperators #-}"
+        ]
+    builtins m
+        | LF.unModuleName (LF.moduleName m) == ["DA", "Internal", "LF"] =
+            [ ""
+            , "data TextMap a = TextMap GHC.Types.Opaque"
+            , "data Time = Time GHC.Types.Opaque"
+            , "data Date = Date GHC.Types.Opaque"
+            , "data ContractId a = ContractId GHC.Types.Opaque"
+            , "data Update a = Update GHC.Types.Opaque"
+            , "data Scenario a = Scenario GHC.Types.Opaque"
+            , "data Party = Party GHC.Types.Opaque"
+            ]
+        | LF.unModuleName (LF.moduleName m) == ["DA", "Internal", "Template"] =
+            [ ""
+            , "class Template c where"
+            , "   signatory :: c -> [DA.Internal.LF.Party]"
+            ]
+        | otherwise = []
+
+newtype DontQualify = DontQualify Bool
```

Comment:
Double negations like `DontQualify False` make my brain explode. :)
We should consider renaming this to `Qualify` and negate all the payloads in the future.